// excalibur@0.29.3/build/dist/EntityComponentSystem/World.js
import { Entity } from './Entity';
import { EntityManager } from './EntityManager';
import { QueryManager } from './QueryManager';
import { System, SystemType } from './System';
import { SystemManager, isSystemConstructor } from './SystemManager';
/**
 * The World is a self-contained entity component system for a particular context.
 */
export class World {
    /**
     * The context type is passed to the system updates
     * @param scene
     */
    constructor(scene) {
        this.scene = scene;
        this.queryManager = new QueryManager(this);
        this.entityManager = new EntityManager(this);
        this.systemManager = new SystemManager(this);
    }
    /**
     * Query the ECS world for entities that match your components
     * @param requiredTypes
     */
    query(requiredTypes) {
        return this.queryManager.createQuery(requiredTypes);
    }
    queryTags(requiredTags) {
        return this.queryManager.createTagQuery(requiredTags);
    }
    /**
     * Update systems by type and time elapsed in milliseconds
     */
    update(type, delta) {
        if (type === SystemType.Update) {
            this.entityManager.updateEntities(this.scene, delta);
        }
        this.systemManager.updateSystems(type, this.scene, delta);
        this.entityManager.findEntitiesForRemoval();
        this.entityManager.processComponentRemovals();
        this.entityManager.processEntityRemovals();
    }
    add(entityOrSystem) {
        if (entityOrSystem instanceof Entity) {
            this.entityManager.addEntity(entityOrSystem);
        }
        if (entityOrSystem instanceof System || isSystemConstructor(entityOrSystem)) {
            this.systemManager.addSystem(entityOrSystem);
        }
    }
    /**
     * Get a system out of the ECS world
     */
    get(system) {
        return this.systemManager.get(system);
    }
    remove(entityOrSystem, deferred = true) {
        if (entityOrSystem instanceof Entity) {
            this.entityManager.removeEntity(entityOrSystem, deferred);
        }
        if (entityOrSystem instanceof System) {
            this.systemManager.removeSystem(entityOrSystem);
        }
    }
    get entities() {
        return this.entityManager.entities;
    }
    clearEntities() {
        this.entityManager.clear();
    }
    clearSystems() {
        this.systemManager.clear();
    }
}
