// excalibur@0.29.3/build/dist/Graphics/Line.js
import { BoundingBox } from '../Collision/BoundingBox';
import { Color } from '../Color';
import { Graphic } from './Graphic';
export class Line extends Graphic {
    constructor(options) {
        super();
        this.color = Color.Black;
        this.thickness = 1;
        const { start, end, color, thickness } = options;
        this.start = start;
        this.end = end;
        this.color = color !== null && color !== void 0 ? color : this.color;
        this.thickness = thickness !== null && thickness !== void 0 ? thickness : this.thickness;
        this._localBounds = this._calculateBounds();
        const { width, height } = this._localBounds;
        this.width = width;
        this.height = height;
    }
    get localBounds() {
        return this._localBounds;
    }
    _calculateBounds() {
        const lineNormal = this.end.sub(this.start).normal();
        const halfThickness = this.thickness / 2;
        const points = [
            this.start.add(lineNormal.scale(halfThickness)),
            this.end.add(lineNormal.scale(halfThickness)),
            this.end.add(lineNormal.scale(-halfThickness)),
            this.start.add(lineNormal.scale(-halfThickness))
        ];
        return BoundingBox.fromPoints(points);
    }
    _drawImage(ctx, _x, _y) {
        ctx.drawLine(this.start, this.end, this.color, this.thickness);
    }
    clone() {
        return new Line({
            start: this.start,
            end: this.end,
            color: this.color,
            thickness: this.thickness
        });
    }
}
