// excalibur@0.29.3/build/dist/Collision/Colliders/Collider.js
import { Vector } from '../../Math/vector';
import { createId } from '../../Id';
import { EventEmitter } from '../../EventEmitter';
/**
 * A collision collider specifies the geometry that can detect when other collision colliders intersect
 * for the purposes of colliding 2 objects in excalibur.
 */
export class Collider {
    constructor() {
        this.id = createId('collider', Collider._ID++);
        /**
         * Composite collider if any this collider is attached to
         *
         * **WARNING** do not tamper with this property
         */
        this.composite = null;
        this.events = new EventEmitter();
        /**
         * Pixel offset of the collision collider relative to the collider, by default (0, 0) meaning the collider is positioned
         * on top of the collider.
         */
        this.offset = Vector.Zero;
    }
    /**
     * Returns a boolean indicating whether this body collided with
     * or was in stationary contact with
     * the body of the other [[Collider]]
     */
    touching(other) {
        const contact = this.collide(other);
        if (contact) {
            return true;
        }
        return false;
    }
}
Collider._ID = 0;
